src/pkg/plug/scanner/api/registration.go

```diff
@@ -0,0 +1,170 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"github.com/goharbor/harbor/src/core/promgr/metamgr"
+	"github.com/goharbor/harbor/src/pkg/plug/scanner"
+	dscanner "github.com/goharbor/harbor/src/pkg/plug/scanner/dao/scanner"
+	"github.com/goharbor/harbor/src/pkg/plug/scanner/scan"
+	"github.com/goharbor/harbor/src/pkg/q"
+	"github.com/pkg/errors"
+)
+
+const (
+	proScannerMetaKey = "projectScanner"
+)
+
+// DefaultController is a singleton api controller for plug scanners
+var DefaultController = New()
+
+// New a basic controller
+func New() Controller {
+	return &basicController{
+		manager:    scanner.New(),
+		proMetaMgr: metamgr.NewDefaultProjectMetadataManager(),
+	}
+}
+
+// basicController is default implementation of api.Controller interface
+type basicController struct {
+	// managers for managing the scanner registrations
+	manager scanner.Manager
+	// for operating the project level configured scanner
+	proMetaMgr metamgr.ProjectMetadataManager
+	// controller for scan actions
+	scan.Controller
+	// Client
+}
+
+// ListRegistrations ...
+func (bc *basicController) ListRegistrations(query *q.Query) ([]*dscanner.Registration, error) {
+	return bc.manager.List(query)
+}
+
+// CreateRegistration ...
+func (bc *basicController) CreateRegistration(registration *dscanner.Registration) (string, error) {
+	// TODO: Get metadata from the adapter service first
+	return bc.manager.Create(registration)
+}
+
+// GetRegistration ...
+func (bc *basicController) GetRegistration(registrationUUID string) (*dscanner.Registration, error) {
+	return bc.manager.Get(registrationUUID)
+}
+
+// RegistrationExists ...
+func (bc *basicController) RegistrationExists(registrationUUID string) bool {
+	registration, err := bc.manager.Get(registrationUUID)
+
+	return !(err == nil && registration == nil)
+}
+
+// UpdateRegistration ...
+func (bc *basicController) UpdateRegistration(registration *dscanner.Registration) error {
+	return bc.manager.Update(registration)
+}
+
+// SetDefaultRegistration ...
+func (bc *basicController) DeleteRegistration(registrationUUID string) (*dscanner.Registration, error) {
+	registration, err := bc.manager.Get(registrationUUID)
```

Comment:
err not handle when get the registration got internal server error

Reply:
ok

Reply:
handled